mycroft/skills/intent/__init__.py

```diff
@@ -31,17 +34,61 @@ class IntentSkill(MycroftSkill):
     def __init__(self):
         MycroftSkill.__init__(self, name="IntentSkill")
         self.engine = IntentDeterminationEngine()
+        self.active_skills = [] # [skill_id , timestamp]
+        self.intent_to_skill = {} # intent:source_skill_id
+        self.converse_timeout = 5 # minutes to prune active_skills
         self.reload_skill = False
 
     def initialize(self):
         self.emitter.on('register_vocab', self.handle_register_vocab)
         self.emitter.on('register_intent', self.handle_register_intent)
         self.emitter.on('recognizer_loop:utterance', self.handle_utterance)
         self.emitter.on('detach_intent', self.handle_detach_intent)
+        self.emitter.on('converse_status_response', self.handle_conversation_response)
+
+    def do_conversation(self, utterances, skill):
+        self.emitter.emit(Message("converse_status_request", {
+                          "skill_id": skill, "utterances": utterances}))
+        self.waiting = True
+        self.result = False
+        while self.waiting:
+            pass
+        return self.result
+
+    def handle_conversation_response(self, message):
+        # id = message.data["skill_id"]
+        # no need to crosscheck id because waiting before new request is made
+        # no other skill will make this request is safe assumption
+        result = message.data["result"]
+        self.result = result
+        self.waiting = False
+
+    def remove_active_skill(self, skill_name):
+        for skill in self.active_skills:
+            if skill[0]==skill_name:
+                self.active_skills.remove(skill)
+
+    def add_active_skill(self, skill_name):
+        # you have to search the list for an existing entry that already contains it and remove that reference
+        self.remove_active_skill(skill_name)
+        # add skill with timestamp to start of skill_list
+        self.active_skills.insert(0, [skill_name, time.time()])
 
     def handle_utterance(self, message):
+
         utterances = message.data.get('utterances', '')
 
+        for skill in self.active_skills:
+            # check for conversation time-out
+            if time.time() - skill[1] >= self.converse_timeout * 60:
+                self.remove_active_skill(skill[0])
```

Comment:
This is going to be a problem.  You never want to alter a list while you are running through it.  You end up changing it "out from under" the loop index and not getting what you expect.  Here is an example that mirrors what you are doing here:

>>> mylist = [1,2,3,4,5,6,7,8]
>>> for x in mylist:
...     if x % 2 == 0:
...         mylist.remove(x)
...     print x
...
1
2
4
6
8
>>> mylist
[1, 3, 5, 7]

Instead, you have to edit the list first, then loop through the edited list.  Something like this:

>>> mylist = [1,2,3,4,5,6,7,8]
>>> mylist = [x for x in mylist if x % 2 == 0]
>>> mylist
[2, 4, 6, 8]

Now you can safely loop through the edited list:

for x in mylist:
    print x